src/alexandria3k/processes/distinguish_affiliations.py

```diff
@@ -0,0 +1,172 @@
+            ColumnMeta("gpt_name"),
+            ColumnMeta("city"),
+            ColumnMeta("ror_id"),
+        ],
+    ),
+]
+
+
+def process(database_path):
+    """
+    This process is used to distinguish the affiliations mentioned in the crossref dataset. It
+    uses the GPT-4 model to extract the affiliation and city from the affiliation text and
+    match it to the ROR database based on levenshtein distance.
+
+    :param database_path: The path to the database
+    """
+    database = apsw.Connection(database_path)
+    client = OpenAI()
+    database.execute(log_sql("DROP TABLE IF EXISTS distinct_affiliations"))
+    database.execute(log_sql(tables[0].table_schema()))
+    set_fast_writing(database)
+    ensure_table_exists(database, "distinct_affiliations")
+    ensure_table_exists(database, "research_organizations")
+
+    select_cursor = database.cursor()
+    select_cursor_2 = database.cursor()
+    insert_cursor = database.cursor()
+
+    distinct_affiliations_number = 0
+    time_start = time.time()
+    # Iterate over all the affiliations mentioned in crossref
+    for mentioned_name in select_cursor.execute(
+        """
+        SELECT DISTINCT name FROM distinct_affiliations
+        """
+    ):
+        affiliation_name = mentioned_name[0]
+        if not mentioned_name:
+            continue
+        # Prompt for the GPT-4 model to extract the affiliation and city from the affiliation text
+        prompt = (
+            "From the given textual piece, identify ONLY the university/research organization and "
+            "city mentioned. The response should be (organization, city). DO NOT produce any "
+            "other textual information. Ignore all other information mentioned in the textual "
+            "piece. If organization or city is not recognized, then return ( , )"
+            + str(mentioned_name)
+        )
+
+        completion = query_response(client, prompt)
+
+        gpt_city, gpt_org = format_result(completion)
+
+        best_ror_id = find_best_ror(gpt_org, select_cursor_2)
+        # Insert the best affiliation match into the distinct_affiliations table (if found)
+        if best_ror_id != -1:
+            insert_cursor.execute(
+                "INSERT INTO distinct_affiliations VALUES (?, ?, ?, ?, ?)",
+                (
+                    distinct_affiliations_number,
+                    affiliation_name,
+                    gpt_org,
+                    gpt_city,
+                    best_ror_id,
+                ),
+            )
+        distinct_affiliations_number += 1
+
+    select_cursor.close()
+    select_cursor_2.close()
+    insert_cursor.close()
+    print(f"llm_process_valid_authors took {time.time() - time_start} seconds")
+    perf.log(
+        f"link_aa_llm added {distinct_affiliations_number} distinct affiliations"
+    )
+
+
+def find_best_ror(gpt_org, select_cursor_2):
+    """
+    This function is used to find the best affiliation match based on levenshtein distance.
```

Comment:
Do not start your comments with "This function…" Just write in imperative voice what the function does. (In all functions.)